bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/util/ModelUtil.java

```diff
@@ -138,24 +141,32 @@ public static void verifyLinkIntermediateEvents(
     // get link throw events definition
     final List<EventDefinition> linkThrowEvents = getEventDefinitionsForLinkThrowEvents(element);
 
-    // get link catch events names
-    final Set<String> linkCatchList =
+    // link catch events names group by scope
+    final Map<BpmnModelElementInstance, Set<String>> linkCatchEventsGroupByScope =
         getEventDefinition(linkCatchEvents, LinkEventDefinition.class)
             .filter(def -> def.getName() != null && !def.getName().isEmpty())
-            .map(LinkEventDefinition::getName)
-            .collect(Collectors.toSet());
+            .collect(
+                groupingBy(
+                    LinkEventDefinition::getScope,
+                    mapping(LinkEventDefinition::getName, Collectors.toSet())));
 
-    // get link throw events names
-    final Set<String> linkThrowList =
+    // link throw events names group by scope
+    final Map<BpmnModelElementInstance, Set<String>> linkThrowEventsGroupByScope =
         getEventDefinition(linkThrowEvents, LinkEventDefinition.class)
             .filter(def -> def.getName() != null && !def.getName().isEmpty())
-            .map(LinkEventDefinition::getName)
-            .collect(Collectors.toSet());
-
-    linkThrowList.forEach(
-        item -> {
-          if (!linkCatchList.contains(item)) {
-            errorCollector.accept(noPairedLinkNames(item));
+            .collect(
+                groupingBy(
+                    LinkEventDefinition::getScope,
+                    mapping(LinkEventDefinition::getName, Collectors.toSet())));
+
+    linkThrowEventsGroupByScope.forEach(
```

Comment:
🔧 Same here, you can extract this validation in a separate private method. You can also add javadoc to explain what this validation is supposed to do.

🔧 Talking about readability, you can change this validation like this (the name of this method can be improved):

```
    private void validateLinkEvents(Map<BpmnModelElementInstance, Set<String>> throwEventsGroupByScope,
                                    Map<BpmnModelElementInstance, Set<String>> catchEventsGroupByScope,
                                    Consumer<String> errorCollector) {
        throwEventsGroupByScope.forEach((scope, throwEventNames) -> {
            Set<String> catchEventNamesForScope = catchEventsGroupByScope.getOrDefault(scope, Collections.emptySet());
            throwEventNames.stream()
                .filter(eventName -> !catchEventNamesForScope.contains(eventName))
                .forEach(eventName -> errorCollector.accept(noPairedLinkNames(eventName)));
        });
    }
```